### PR TITLE
ch4/ofi: Add psm3 capability set

### PR DESCRIPTION
**Pull request by nitbhat**
## Pull Request Description
This PR adds a capability set for the PSM3 provider. All of the current configuration is borrowed from the PSM2 provider.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
We disabled libfabric psm3 -- https://github.com/pmodels/mpich/blob/ebe7e0a480fe87edc90f4d5bddcaacca34e286e0/src/mpid/ch4/netmod/ofi/subconfigure.m4#L276 -- due to compile issues on our Jenkins. We need to revisit the issue and see if it can be resolved.

**Comment by nitbhat** [comment]
@hzhou, shall I enable psm3 and test this on jenkins to see what compile issues you're getting?

**Comment by hzhou** [comment]
> @hzhou, shall I enable psm3 and test this on jenkins to see what compile issues you're getting?

Sure. Go ahead

**Comment by hzhou** [comment]
@nitbhat When you done, use the custom test for it. For example -- https://github.com/pmodels/mpich/pull/5904#issuecomment-1076868034

The default review tests will use pre-built modules, which skips the rebuild of libfabric.

**Comment by nitbhat** [comment]
test:mpich/custom
netmod: ch4:ofi

**Comment by nitbhat** [comment]
test:mpich/custom
netmod: ch4:ofi

**Comment by hzhou** [comment]
Looks like it is working now. For reference, the original reason we disabled `psm3` is here -- https://github.com/pmodels/mpich/pull/5344#issuecomment-856838489

**Comment by nitbhat** [comment]
@hzhou: The build looks good and the tests seem to be working. But, is there a way I can specify to the test framework on jenkins to use the psm3 provider? (maybe with netmod:ch4:ofi:psm3?)

**Comment by hzhou** [comment]
> @hzhou: The build looks good and the tests seem to be working. But, is there a way I can specify to the test framework on jenkins to use the psm3 provider? (maybe with netmod:ch4:ofi:psm3?)
> 
> I see that sockets is used as the provider for the tests.

Our Jenkins only have Infiniband card. Does `psm3` work with that?

**Comment by nitbhat** [comment]
Yes, psm3 works with Infiniband, it is a performant alternative to sockets over IB.

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
env: FI_PROVIDER=psm3

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by nitbhat** [comment]
test:mpich/custom
netmod: ch4:ofi
env: FI_PROVIDER=psm3

**Comment by nitbhat** [comment]
There is only one failure seen for psm3. I have seen this issue before and it happens intermittently for some tests. I've opened https://github.com/pmodels/mpich/issues/5975 and will add an xfail to this PR.

File: test/mpi/maint/jenkins/xfail.conf [hunk 1]
```diff
@@ -86,6 +86,9 @@
 # Sunf90 forbid passing cray pointer as integer
 * solstudio * * *       /^allocmemf90/          xfail=ticket0      f90/ext/testlist
 
+# psm3 specific failures
+* * * ch4:ofi:psm3 *    /^p_red .*/                      xfail=issue5975        coll/testlist
```
Comment (nitbhat) [inline test/mpi/maint/jenkins/xfail.conf:90]: @hzhou: Is the right way to xfail a psm3 specific failure?

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi

**Comment by nitbhat** [comment]
> > There is only one failure seen for psm3. I have seen this issue before and it happens intermittently for some tests. I've opened #5975 and will add an xfail to this PR.
> 
> The xfail entry won't work against the custom tests since `psm3` is selected via `FI_PROVIDER`, not configure device option. Since we don't have review tests targeting `psm3` yet, the xfail entries is not necessary. Once we add such jobs, we can add xfail entries targeting the particular job or jenkins config option. Hopefully, the bug is fixed by then 🤞

Okay, understood. I'll go ahead and remove that commit in that case.

**Comment by nitbhat** [comment]
test:mpich/custom
netmod: ch4:ofi

**Comment by hzhou** [comment]
@nitbhat Does `psm3` makes `psm2` obsolete?

**Comment by nitbhat** [comment]
> @nitbhat Does `psm3` make `psm2` obsolete?

@hzhou: No, psm2 continues to be the preferred provider for OPA.

**Review by hzhou — APPROVED** [review]
LGTM

**Comment by hzhou** [comment]
@nitbhat Approved. Could you rebase?

**Comment by nitbhat** [comment]
> @nitbhat Approved. Could you rebase?
Yes, rebased.

**Comment by hzhou** [comment]
test:mpich/warnings

**Pull request merged**